test/test_helpers.py

```diff
@@ -74,6 +76,101 @@ def fake_script_calls(test_case: unittest.TestCase,
     return calls  # type: ignore
 
 
+def create_framework(
+        request: pytest.FixtureRequest,
+        *,
+        meta: typing.Optional[ops.CharmMeta] = None):
+    env_backup = os.environ.copy()
+    os.environ['PATH'] = os.pathsep.join([
+        str(pathlib.Path(__file__).parent / 'bin'),
+        os.environ['PATH']])
+    os.environ['JUJU_UNIT_NAME'] = 'local/0'
+
+    tmpdir = pathlib.Path(tempfile.mkdtemp())
+
+    class CustomEvent(ops.EventBase):
+        pass
+
+    class TestCharmEvents(ops.CharmEvents):
+        custom = ops.EventSource(CustomEvent)
+
+    # Relations events are defined dynamically and modify the class attributes.
+    # We use a subclass temporarily to prevent these side effects from leaking.
+    ops.CharmBase.on = TestCharmEvents()  # type: ignore
+
+    if meta is None:
+        meta = ops.CharmMeta()
+    model = ops.Model(meta, _ModelBackend('local/0'))  # type: ignore
+    # We can pass foo_event as event_name because we're not actually testing dispatch.
+    framework = ops.Framework(SQLiteStorage(':memory:'), tmpdir, meta, model)  # type: ignore
+
+    def finalizer():
+        os.environ.clear()
+        os.environ.update(env_backup)
+        shutil.rmtree(tmpdir)
+        ops.CharmBase.on = ops.CharmEvents()  # type: ignore
+        framework.close()
+
+    request.addfinalizer(finalizer)
+
+    return framework
+
+
+class FakeScript:
+    def __init__(self,
+                 request: pytest.FixtureRequest,
+                 path: typing.Optional[pathlib.Path] = None):
+        if path is None:
+            fake_script_path = tempfile.mkdtemp('-fake_script')
+            self.path = pathlib.Path(fake_script_path)
+            old_path = os.environ["PATH"]
+            os.environ['PATH'] = os.pathsep.join([fake_script_path, os.environ["PATH"]])
```

Comment:
Do we not need to modify `os.environ` if `path` is passed?

Reply:
I was wondering the same when I saw this (got this logic from the original function), but after running all tests, it seems we don't need to do so.